independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BuiltinBean.java

```diff
@@ -449,117 +464,118 @@ private static ResultHandle loadInvokerTargetBean(InvokerInfo invoker, BytecodeC
+        if (ctx.injectionTarget.kind() != TargetKind.OBSERVER) {
+            ctx.errors.accept(new DefinitionException(
+                    "EventMetadata can be only injected into an observer method: " + ctx.injectionPoint.getTargetInfo()));
         }
     }
 
-    private static void validateInterceptionProxy(InjectionTargetInfo injectionTarget,
-            InjectionPointInfo injectionPoint, Consumer<Throwable> errors) {
-        if (injectionTarget.kind() != TargetKind.BEAN
-                || (!injectionTarget.asBean().isProducerMethod() && !injectionTarget.asBean().isSynthetic())
-                || injectionTarget.asBean().getInterceptionProxy() == null) {
-            errors.accept(new DefinitionException(
+    private static void validateInterceptionProxy(ValidatorContext ctx) {
+        if (ctx.injectionTarget.kind() != TargetKind.BEAN
+                || (!ctx.injectionTarget.asBean().isProducerMethod() && !ctx.injectionTarget.asBean().isSynthetic())
+                || ctx.injectionTarget.asBean().getInterceptionProxy() == null) {
+            ctx.errors.accept(new DefinitionException(
                     "InterceptionProxy can only be injected into a producer method or a synthetic bean"));
         }
-        if (injectionPoint.getType().kind() != Kind.PARAMETERIZED_TYPE) {
-            errors.accept(new DefinitionException("InterceptionProxy must be a parameterized type"));
+        if (ctx.injectionPoint.getType().kind() != Kind.PARAMETERIZED_TYPE) {
+            ctx.errors.accept(new DefinitionException("InterceptionProxy must be a parameterized type"));
         }
-        Type interceptionProxyType = injectionPoint.getType().asParameterizedType().arguments().get(0);
+        Type interceptionProxyType = ctx.injectionPoint.getType().asParameterizedType().arguments().get(0);
         if (interceptionProxyType.kind() != Kind.CLASS && interceptionProxyType.kind() != Kind.PARAMETERIZED_TYPE) {
-            errors.accept(new DefinitionException(
+            ctx.errors.accept(new DefinitionException(
                     "Type argument of InterceptionProxy may only be a class or parameterized type"));
         }
-        if (!injectionTarget.asBean().getProviderType().equals(interceptionProxyType)) {
-            String msg = injectionTarget.asBean().isProducerMethod()
+        if (!ctx.injectionTarget.asBean().getProviderType().equals(interceptionProxyType)) {
+            String msg = ctx.injectionTarget.asBean().isProducerMethod()
                     ? "Type argument of InterceptionProxy must be equal to the return type of the producer method"
                     : "Type argument of InterceptionProxy must be equal to the bean provider type";
-            errors.accept(new DefinitionException(msg));
+            ctx.errors.accept(new DefinitionException(msg));
+        }
+        ClassInfo clazz = getClassByName(ctx.beanDeployment.getBeanArchiveIndex(), interceptionProxyType.name());
```

Comment:
Since we're sure that `injectionTarget.kind() == TargetKind.BEAN` at this point you don't need an explicit `BeanDeployment` reference but use `injectionTarget.asBean().getDeployment().getBeanArchiveIndex()` instead.

Reply:
Good point. It's a mouthful, but would let me avoid the refactoring. Which, on the other hand, makes validation better in my eyes...

Reply:
Kept this, because it makes the code shorter :-)